### PR TITLE
[mempool] add configurable default_failover

### PR DESCRIPTION
**Pull request by christinacelee**
## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/libra/developers.libra.org, and link to your PR here.)

### CODE REVIEW

File: mempool/src/shared_mempool/peer_manager.rs [hunk 1]
```diff
@@ -522,41 +522,19 @@ impl PeerManager {
...
 
-    // checks whether a peer is a chosen broadcast recipient:
-    // - all primary peers
-    // - fallback peers, if k-policy is enabled
+    // Broadcast to ALL upstream peers, for better reliability of the txn reaching a validator
     // this does NOT check for whether this peer is alive
+    // In the future, we can consider mempool communicating with state sync to detect whether certain
+    // upstream peers are up-to-date enough to broadcast txns to (otherwise the txns might end up not
+    // being ready for broadcast on the upstream peer), but for now we will resort to broadcasting
+    // to all peers across all upstream networks
```
Comment (davidiw) [inline mempool/src/shared_mempool/peer_manager.rs:534]: might be good to write up an issue with your thoughts on state sync / mempool interrelationship (or a quip)

**Review by davidiw — APPROVED** [review]
I think this is actually good for now -- we won't have to worry about a down upstream causing sla issues, but we'll definitely need to revisit this as we plan the future of mempool.

File: mempool/src/shared_mempool/peer_manager.rs [hunk 2]
```diff
@@ -143,6 +147,11 @@ impl PeerManager {
                 .dec();
             state.is_alive = false;
         }
+
+        {
+            let mut blah = self.default_failovers.lock();
```
Comment (davidiw) [inline mempool/src/shared_mempool/peer_manager.rs:152]: Naming? Also can you make this one line?

File: mempool/src/shared_mempool/peer_manager.rs [hunk 3]
```diff
@@ -349,6 +359,30 @@ impl PeerManager {
...
+        // broadcast/consensus on that end based on its stale state.
+        // So:
+        // (1) don't invest too much in refactoring this code w.r.t. the rest of this function
+        // (2) the logic of this function should later be migrated to a smarter networking layer
+        // that can handle peer selection for mempool
+        let mut default_failovers = self.default_failovers.lock();
+        if default_failovers.len() < self.mempool_config.default_failovers {
+            for failover_network in self.upstream_config.networks[1..].iter() {
+                if let Some(active_peers) = active_peers_by_network.get(failover_network) {
+                    for p in active_peers.iter().cloned() {
```
Comment (davidiw) [inline mempool/src/shared_mempool/peer_manager.rs:375]: Redundant clone?

**Review by davidiw — APPROVED** [review]
Can we get some metrics here? Especially around time for sending during fail over with and without?

File: mempool/src/shared_mempool/peer_manager.rs (hunk 3 above)
Comment (christinacelee) [reply mempool/src/shared_mempool/peer_manager.rs:375]: apparently this is a double reference, so it complains when I do a double clone inside the loop... :/

**Pull request closed without merging**